### ISSUE TITLE
array_flip(): Can only flip STRING and INTEGER values! output.inc:184

### ISSUE THREAD

**Issue by bradjones1** [drush-ops/drush#1667]
```
bradj@home:~/www/brjllc$ ddrush uli
array_flip(): Can only flip STRING and INTEGER values! output.inc:184
```
Call trace from Xdebug:
![screenshot from 2015-10-05 15 51 52](https://cloud.githubusercontent.com/assets/981966/10294689/26db7bca-6b79-11e5-8b04-a95dee2c12f2.png)

Documenting for the moment since I don't see a related ticket on search... will try to debug later.

**Comment by weitzman** [comment]
we'll need steps to reproduce.

which xdebug client is that?

**Comment by jonhattan** [comment]
I'm seeing this when `drush @site uli`. 

```
Array
(
    [!phase] => 5
    [!max] => 
)
array_flip(): Can only flip STRING and INTEGER values! output.inc:185     [warning]
```

It comes from `drush_log(dt("Find command files for phase !phase (max=!max)", array('!phase' => $phase, '!max' => $phase_max)), 'debug');` when `$phase_max` is FALSE.

**Comment by jonhattan** [comment]
Fixed in master and 7.x

**Comment by FGM** [comment]
Same error on master Drush nad master D8 (just pulled both right now) when creating a command with required-arguments truthy, and not passing the required argument. The error appears to occur because the $args passed to find_legacy_dt_args() is like : ['@missing' => ..., '@required' => ..., '@command' => ...] while that function expected !-prefixed strings, not @-prefixed ones.

**Comment by weitzman** [comment]
Just tried on master - works fine for me.

**Comment by FGM** [comment]
Indeed, with today's HEAD on drush and core, the symptom changes in that case : it is now `Illegal offset type drush.inc:1584` which is a different issue, happening because in `drush_set_error()`, `$error` is not a string but a TranslatableMarkup instance, causing `$error_log[$error][] = $message;` to emit this warning, which can be removed by wrapping $error in double quotes to trigger its `__toString()`.

**Comment by weitzman** [comment]
The doxygen for drush_set_error() needs clarity on what types $error can be. Mind to send a PR? Thanks for retesting.

**Comment by FGM** [comment]
I can do this, but AIUI the error is still in Drush code, which calls drush_set_error() incorrectly. Here is a drush plugin to reproduce the problem:

```php
function bug_drush_command() {
  return [
    'foo' => [
      'arguments' => ['arg' => 'bar'],
      'required-arguments' => TRUE,
    ],
  ];
}

function bug_foo($arg) {
  drush_print_r($arg);
}
```
Invoking gives this:
```
$  drush foo
Illegal offset type drush.inc:1584                                                                                 [warning]
Missing required argument: arg.  See `drush help foo` for information on usage.  
```
